edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs

```diff
@@ -238,9 +249,57 @@ async void OnConnectionStatusChanged(ConnectionStatus status, ConnectionStatusCh
 
                 if (this.pullOnReconnect && this.initTask.IsCompleted && status == ConnectionStatus.Connected)
                 {
+                    var delayedTwinPull = true;
                     using (await this.twinLock.LockAsync())
                     {
-                        await this.RefreshTwinAsync();
+                        var now = DateTime.Now;
+                        if (now - this.lastTwinPullOnConnect > this.TwinPullOnConnectThrottleTime && !this.isDelayedTwinPullInProgress.Get())
+                        {
+                            this.lastTwinPullOnConnect = now;
+                            await this.RefreshTwinAsync();
+                            delayedTwinPull = false;
+                        }
+                    }
+
+                    if (delayedTwinPull)
+                    {
+                        if (this.isDelayedTwinPullInProgress.GetAndSet(true))
+                        {
+                            this.pullRequestCounter++;
+                        }
+                        else
+                        {
+                            _ = Task.Run(
+                                async () =>
+                                {
+                                    await Task.Delay(this.TwinPullOnConnectThrottleTime);
```

Comment:
Why do we delay for 30s? Is there a certain max RPS that we calculated this from or is just a reasonable number?

Reply:
It is just that "120 requests per hour if something goes wrong is not that bad". Also if the pull gets delayed unreasonably (due to a program error), then 30 sec delay is not horrible. I did not want to make it too short, so it will not pollute the log with messages, but also did not want to make it too long, so the 'blind window" (when desired properties are updated but the agent does not know about it) will not be too big. I though something around 30sec-60sec is a good number